pkg/cgroup/v1/set.go

```diff
@@ -0,0 +1,144 @@
+/*
+Copyright 2021 Andy Dalton
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cgroup
+
+import (
+	"fmt"
+	"log"
+	"strings"
+
+	"github.com/adalton/teleport-exercise/pkg/adaptation/os"
+
+	"github.com/google/uuid"
+)
+
+const (
+	DefaultBasePath                   = "/sys/fs/cgroup"
+	defaultDirectoryPerms os.FileMode = 0755
+)
+
+// Set maintains a collection of 0 or more cgroup controllers that should be
+// created/removed at the same time.
+type Set struct {
+	osAdapter   *os.Adapter
+	basePath    string
+	jobId       uuid.UUID
+	controllers []Controller
+}
+
+// NewSet creates a new cgroup (v1) set for the given jobID.  This assumes that
+// the cgroup filesystem is mounted at /sys/fs/cgroup.
+func NewSet(jobId uuid.UUID, controllers ...Controller) *Set {
+	return NewSetDetailed(nil, DefaultBasePath, jobId, controllers...)
+}
+
+// NewSetDetailed creates a new cgroup (v1) set for the given jobID rooted
+// at the given basePath.
+func NewSetDetailed(
+	osAdapter *os.Adapter,
+	basePath string,
+	jobId uuid.UUID,
+	controllers ...Controller,
+) *Set {
+
+	return &Set{
+		osAdapter:   osAdapter,
+		basePath:    basePath,
+		jobId:       jobId,
+		controllers: controllers,
+	}
+}
+
+// Create creates the cgroup v1 directories for all registered controllers.
+func (s *Set) Create() (retErr error) {
+	if s == nil {
+		// If the set is nil, then Create is vacuously successful
+		return nil
+	}
+
+	failPoint := -1
+
+	for i := range s.controllers {
+		path := fmt.Sprintf("%s/%s/jobs/%s", s.basePath, s.controllers[i].Name(), s.jobId.String())
```

Comment:
Is it worth refactoring this to a helper method? It appears a few times throughout the code.